# Stockholm
_namespace: [SMRUCC.genomics.Data.Xfam.Rfam](./index.md)_

Rfam.seed
 
 Stockholm format is a Multiple sequence alignment format used by Pfam and Rfam to disseminate protein and RNA sequence alignments. 
 The alignment editors Ralee and Belvu support Stockholm format as do the probabilistic database search tools, 
 Infernal and HMMER, and the phylogenetic analysis tool Xrate.




### Properties

#### AC
Accession number: 
 Accession number In form PFxxxxx (Pfam) Or RFxxxxx (Rfam).
#### AU
Author 
 Authors of the entry.
#### BM
Build method 
 Command line used To generate the model
#### CB
Calibration method: 
 Command line used To calibrate the model (Rfam only, release 12.0 And later)
#### CC
Comment : 
 Comments.
#### CL
Clan: 
 Clan accession
#### DC
Database Comment: 
 Comment about database reference.
#### DE
Definition 
 Short description of family.
#### DR
Database Reference: 
 Reference to external database.
#### FR
False discovery Rate: 
 A method used To Set the bit score threshold based On the ratio Of expected false positives to true positives. Floating point number between 0 And 1.
#### GA
Gathering threshold 
 Search threshold To build the full alignment.
#### ID
Identification: 
 One word name For family.
#### KW
Keywords : 
 Keywords.
#### MB
Membership : 
 Used for listing Clan membership
#### NC
Noise Cutoff 
 Highest sequence score (And domain score For Pfam) Of match Not In full alignment.
#### NE
Pfam accession: 
 Indicates a nested domain.
#### NH
New Hampshire 
 A tree In New Hampshire eXtended format.
#### NL
Location : 
 Location of nested domains - sequence ID, start And end of insert.
#### PI
Previous identifier: 
 Record of all previous ID lines.
#### RA
Reference Author: 
 Reference Author
#### RC
Reference Comment: 
 Comment about literature reference.
#### RL
Reference Location: 
 Journal location.
#### RM
Reference Medline: 
 Eight digit medline UI number.
#### RN
Reference Number: 
 Reference Number.
#### RT
Reference Title: 
 Reference Title.
#### SE
Source Of seed 
 The source suggesting the seed members belong To one family.
#### SM
Search method 
 Command line used To perform the search
#### SQ
Sequence 
 Number of sequences in alignment.
#### SS
Source Of Structure 
 The source(prediction Or publication) Of the consensus RNA secondary Structure used by Rfam.
#### TC
Trusted Cutoff 
 Lowest sequence score (And domain score For Pfam) Of match In the full alignment.
#### TN
Tree ID 
 A unique identifier For the Next tree.
#### TP
Type 
 Type of family -- presently Family, Domain, Motif Or Repeat for Pfam. -- a tree with roots Gene, Intron Or Cis-reg for Rfam.
#### WK
Wikipedia link: 
 Wikipedia page
