<!DOCTYPE html><html lang="zh-CN">
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, minimum-scale=1.0, maximum-scale=1.0" />
    <title>geneExpression</title>
    <meta name="author" content="xie.guigang@gcmodeller.org" />
    <meta name="copyright" content="SMRUCC genomics Copyright (c) 2020" />
    <meta name="keywords" content="GCModeller; Xanthomonas; Artificial Life" />
    <meta name="generator" content="https://github.com/xieguigang/xDoc" />
    <meta name="theme-color" content="#333" />
    <meta name="last-update" content="2022-11-23" />
    <meta name="description" content="" />
    <meta class="foundation-data-attribute-namespace" />
    <meta class="foundation-mq-xxlarge" />
    <meta class="foundation-mq-xlarge" />
    <meta class="foundation-mq-large" />
    <meta class="foundation-mq-medium" />
    <meta class="foundation-mq-small" />
    <meta class="foundation-mq-topbar" />
  </head>
  <body>
    <table width="100%" summary="page for {geneExpression}">
      <tbody>
        <tr>
          <td>{geneExpression}</td>
          <td style="text-align: right;">R# Documentation</td>
        </tr>
      </tbody>
    </table>
    <h1>geneExpression</h1>
    <hr />
    <p>
      <code>
        <span style="color: blue;">imports</span>
        <span style="color: brown"> "geneExpression"</span>
        <span style="color: blue;"> from</span>
        <span style="color: brown"> "phenotype_kit"</span>
      </code>
    </p>
    <p></p>
    <div id="main-wrapper">
      <table>
        <thead>
          <tr>
            <th>function</th>
            <th>description</th>
          </tr>
        </thead>
        <tbody><tr>
  <td id="tr">
    <a href="./geneExpression/tr.html">tr</a>
  </td>
  <td><p>do matrix transpose</p></td>
</tr>
<tr>
  <td id="dims">
    <a href="./geneExpression/dims.html">dims</a>
  </td>
  <td><p>get summary information about the HTS matrix dimensions</p></td>
</tr>
<tr>
  <td id="as.expr_list">
    <a href="./geneExpression/as.expr_list.html">as.expr_list</a>
  </td>
  <td><p>convert the matrix into row gene list</p></td>
</tr>
<tr>
  <td id="setTag">
    <a href="./geneExpression/setTag.html">setTag</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="setZero">
    <a href="./geneExpression/setZero.html">setZero</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="setFeatures">
    <a href="./geneExpression/setFeatures.html">setFeatures</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="filterZeroSamples">
    <a href="./geneExpression/filterZeroSamples.html">filterZeroSamples</a>
  </td>
  <td><p>filter out all samples columns which its expression vector is ZERO!</p></td>
</tr>
<tr>
  <td id="filterNaNMissing">
    <a href="./geneExpression/filterNaNMissing.html">filterNaNMissing</a>
  </td>
  <td><p>set the NaN missing value to default value</p></td>
</tr>
<tr>
  <td id="load.expr">
    <a href="./geneExpression/load.expr.html">load.expr</a>
  </td>
  <td><p>load an expressin matrix data</p></td>
</tr>
<tr>
  <td id="load.expr0">
    <a href="./geneExpression/load.expr0.html">load.expr0</a>
  </td>
  <td><p>read the binary matrix data file</p></td>
</tr>
<tr>
  <td id="write.expr_matrix">
    <a href="./geneExpression/write.expr_matrix.html">write.expr_matrix</a>
  </td>
  <td><p>write the gene expression data matrix file</p></td>
</tr>
<tr>
  <td id="filter">
    <a href="./geneExpression/filter.html">filter</a>
  </td>
  <td><p>Filter the geneID rows</p></td>
</tr>
<tr>
  <td id="as.generic">
    <a href="./geneExpression/as.generic.html">as.generic</a>
  </td>
  <td><p>cast the HTS matrix object to the general dataset</p></td>
</tr>
<tr>
  <td id="average">
    <a href="./geneExpression/average.html">average</a>
  </td>
  <td><p>calculate average value of the gene expression for<br />
 each sample group.</p>

<p>this method can be apply for reduce data size when <br />
 create some plot for visualize the gene expression<br />
 patterns across the sample groups.</p></td>
</tr>
<tr>
  <td id="z_score">
    <a href="./geneExpression/z_score.html">z_score</a>
  </td>
  <td><p>To avoid the influence of expression level to the <br />
 clustering analysis, z-score transformation can <br />
 be applied to covert the expression values to <br />
 z-scores by performing the following formula:</p>

<p></p>

<pre><code class="">z = (x - u) / sd
</code></pre>

<p>x is value to be converted (e.g., a expression value <br />
 of a genomic feature in one condition), µ is the <br />
 population mean (e.g., average expression value Of <br />
 a genomic feature In different conditions), σ Is the <br />
 standard deviation (e.g., standard deviation of <br />
 expression of a genomic feature in different conditions).</p></td>
</tr>
<tr>
  <td id="pca">
    <a href="./geneExpression/pca.html">pca</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="totalSumNorm">
    <a href="./geneExpression/totalSumNorm.html">totalSumNorm</a>
  </td>
  <td><p>normalize data by sample column</p></td>
</tr>
<tr>
  <td id="relative">
    <a href="./geneExpression/relative.html">relative</a>
  </td>
  <td><p>normalize data by feature rows</p></td>
</tr>
<tr>
  <td id="expression.cmeans_pattern">
    <a href="./geneExpression/expression.cmeans_pattern.html">expression.cmeans_pattern</a>
  </td>
  <td><p>This function performs clustering analysis of time course data. <br />
 Calculate gene expression pattern by cmeans algorithm.</p></td>
</tr>
<tr>
  <td id="expression.cmeans3D">
    <a href="./geneExpression/expression.cmeans3D.html">expression.cmeans3D</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="savePattern">
    <a href="./geneExpression/savePattern.html">savePattern</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="readPattern">
    <a href="./geneExpression/readPattern.html">readPattern</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="cmeans_matrix">
    <a href="./geneExpression/cmeans_matrix.html">cmeans_matrix</a>
  </td>
  <td><p>get cluster membership matrix</p></td>
</tr>
<tr>
  <td id="split.cmeans_clusters">
    <a href="./geneExpression/split.cmeans_clusters.html">split.cmeans_clusters</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="peakCMeans">
    <a href="./geneExpression/peakCMeans.html">peakCMeans</a>
  </td>
  <td><p>This function performs clustering analysis of time course data</p></td>
</tr>
<tr>
  <td id="deg.t.test">
    <a href="./geneExpression/deg.t.test.html">deg.t.test</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="geneId">
    <a href="./geneExpression/geneId.html">geneId</a>
  </td>
  <td><p>get gene Id list</p></td>
</tr>
<tr>
  <td id="as.deg">
    <a href="./geneExpression/as.deg.html">as.deg</a>
  </td>
  <td><p>create gene expression DEG model</p></td>
</tr>
<tr>
  <td id="deg.class">
    <a href="./geneExpression/deg.class.html">deg.class</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="joinSample">
    <a href="./geneExpression/joinSample.html">joinSample</a>
  </td>
  <td><p>do matrix join by samples</p></td>
</tr></tbody>
      </table>
    </div>
  </body>
</html>