<!DOCTYPE html>
<html>

<head>
    <title>cast the cluster data as the enrichment background</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>as.background {background}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>cast the cluster data as the enrichment background</h2>

        <h3>Description</h3>

        

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>as.background</strong>(<i>clusters</i>,
    <i>background.size</i> = <span style='color: green;'>-1</span>,
    <i>name</i> = <span style='color: brown;'><strong>"n/a"</strong></span>,
    <i>tax.id</i> = <span style='color: brown;'><strong>"n/a"</strong></span>,
    <i>desc</i> = <span style='color: brown;'><strong>"n/a"</strong></span>,
    <i>omics</i> = Transcriptomics,
    <i>filter.compoundId</i> = <span style='color: blue !important;'>TRUE</span>,
    <i>kegg.code</i> = <span style='color: blue !important;'>NULL</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>clusters</dt>
<dd><p>a data cluster or a collection of kegg pathway model</p></dd>


<dt>background.size</dt>
<dd><p>default value -1 or zero means auto evaluated. [as integer]</p></dd>


<dt>name</dt>
<dd><p>the background model name. [as string]</p></dd>


<dt>tax.id</dt>
<dd><p>ncbi taxonomy id of the target organism. [as string]</p></dd>


<dt>desc</dt>
<dd><p>the model description. [as string]</p></dd>


<dt>omics</dt>
<dd><p>Create a enrichment background model for run multiple omics data analysis?
 this parameter is only works for the kegg pathway model where you are 
 speicifc via the <strong>`clusters`</strong> parameter. [as <a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/GSEA/OmicsData.html">OmicsData</a>]</p></dd>


<dt>filter.compoundId</dt>
<dd><p>do compound id filtering when target model is <strong>`omics`</strong>?
 (all of the KEGG drug id and KEGG glycan id will be removed from the cluster model). [as boolean]</p></dd>


<dt>kegg.code</dt>
<dd><p>the kegg organism code when the given <strong>`clusters`</strong> collection is
 a collection of the pathway object. [as string]</p></dd>


<dt>env</dt>
<dd><p>[as <a href="/vignettes/clr/SMRUCC/Rsharp/Runtime/Environment.html">Environment</a>]</p></dd>

        </table>


        <h3 style="display: none;">Details</h3>

        

        <h3 style="display: block">Authors</h3>

        gseakit

        <h3>Value</h3>

        this function returns data object of type <a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/GSEA/Background.html">Background</a>.<h4>clr value class</h4><ul><li><a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/GSEA/Background.html">Background</a></li></ul>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>background</em> version 1.0.0.0 <a
                href="../background.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>
