<!DOCTYPE html>
<html>

<head>
    <title>Constructs k-mer adjacency graphs from protein sequence data. Nodes represent k-length</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>kmer_graph {proteinKit}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>Constructs k-mer adjacency graphs from protein sequence data. Nodes represent k-length</h2>

        <h3>Description</h3>

        <p>subsequences, edges connect k-mers appearing consecutively in the sequence.</p>

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>kmer_graph</strong>(<i>prot</i>,
    <i>k</i> = <span style='color: green;'>3</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>prot</dt>
<dd><p>A FASTA sequence or collection of FASTA sequences to process.</p></dd>


<dt>k</dt>
<dd><p>The subsequence length parameter for k-mer generation. Default is 3. [as integer]</p></dd>


<dt>env</dt>
<dd><p>The R runtime environment for error handling and resource cleanup. [as <a href="/vignettes/clr/SMRUCC/Rsharp/Runtime/Environment.html">Environment</a>]</p></dd>

        </table>


        <h3 style="display: none;">Details</h3>

        

        <h3 style="display: block">Authors</h3>

        seqtoolkit

        <h3>Value</h3>

        <p>Returns a single <a href="/vignettes/clr/SMRUCC/genomics/Model/MotifGraph/ProteinStructure/Kmer/KMerGraph.html">KMerGraph</a> for single sequence input. Returns a named list 
 of KMerGraph objects for multiple sequences. Returns error message for invalid inputs.</p><h4>clr value class</h4><ul><li><a href="/vignettes/clr/SMRUCC/genomics/Model/MotifGraph/ProteinStructure/Kmer/KMerGraph.html">KMerGraph</a></li></ul>

        <h3 style="display: block;">Examples</h3>

        <pre><code id="example_r"> let graphs = kmer_graph(sequences, k = 3);</code></pre>

        <hr />
        <div style="text-align: center;">[Package <em>proteinKit</em> version 1.0.0.0 <a
                href="../proteinKit.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>