<!DOCTYPE html>
<html>

<head>
    <title>Calculate the morgan fingerprint based on the k-mer graph data</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>kmer_fingerprint {proteinKit}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>Calculate the morgan fingerprint based on the k-mer graph data</h2>

        <h3>Description</h3>

        <p><br /> Generates fixed-length molecular fingerprint vectors from k-mer graphs using <br /> Morgan algorithm with circular topology hashing.</p>

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>kmer_fingerprint</strong>(<i>graph</i>,
    <i>radius</i> = <span style='color: green;'>3</span>,
    <i>len</i> = <span style='color: green;'>4096</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>graph</dt>
<dd><p>The k-mer graph object to fingerprint. [as <a href="/vignettes/clr/SMRUCC/genomics/Model/MotifGraph/ProteinStructure/Kmer/KMerGraph.html">KMerGraph</a>]</p></dd>


<dt>radius</dt>
<dd><p>Neighborhood radius for structural feature capture. Larger values 
 consider more distant node relationships. Default is 3. [as integer]</p></dd>


<dt>len</dt>
<dd><p>Output vector length (uses modulo hashing). Default 4096. [as integer]</p></dd>

        </table>


        <h3 style="display: none;">Details</h3>

        

        <h3 style="display: block">Authors</h3>

        seqtoolkit

        <h3>Value</h3>

        <p>Integer array fingerprint where indices represent structural features</p><h4>clr value class</h4><ul><li><i>any</i> kind</li></ul>

        <h3 style="display: block;">Examples</h3>

        <pre><code id="example_r"> let seqs = read.fasta("./proteins.fa");
 # create protein sequence graph based on k-mer(k=3)
 let g = kmer_graph(seqs, k = 3);
 
 for(let kmer in g) {
     print(kmer |> kmer_fingerprint());
 }</code></pre>

        <hr />
        <div style="text-align: center;">[Package <em>proteinKit</em> version 1.0.0.0 <a
                href="../proteinKit.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>