<!DOCTYPE html>
<html>

<head>
    <title>Fetch and Download Data for a Specific KO Entry</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="../../_assets/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>fetch_ko_data {kegg_api}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>Fetch and Download Data for a Specific KO Entry</h2>

        <h3>Description</h3>

        <p>An internal function that downloads gene ID associations for a specific
KEGG Orthology (KO) identifier and optionally filters them by species.</p>

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>fetch_ko_data</strong>(<i>db</i>, <i>ko.id</i>, <i>species</i>,
    <i>seqtype</i> = 'c'('ntseq', 'aaseq'),
    <i>download.seqs</i> = <span style='color: blue !important;'>TRUE</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>db</dt>
<dd><p><p>A local repository connection to the target KEGG genes database.</p></p></dd>


<dt>ko.id</dt>
<dd><p><p>A character string specifying the KEGG Orthology identifier
(e.g., \code{"K00001"}).</p></p></dd>


<dt>species</dt>
<dd><p><p>A character vector of KEGG organism species codes for filtering.
Default is \code{NULL} (no filtering). Only genes from the specified organisms
will be downloaded.</p></p></dd>


<dt>seqtype</dt>
<dd><p><p>The type of FASTA sequence data to download. Must be either
\code{"ntseq"} (gene nucleotide sequence) or \code{"aaseq"} (protein amino
acid sequence). Default is \code{"ntseq"}.</p></p></dd>

        </table>


        <h3 style="display: block;">Details</h3>

        <p>This function performs the following steps:
\enumerate{
\item Checks if a cached gene ID list exists for the KO entry
\item If not cached, downloads the gene ID list from the KEGG REST API
\item Parses the gene IDs into species code and gene locus tag components
\item Filters genes by the specified species (if provided)
\item Triggers the download of FASTA sequences for the filtered genes
}
The KEGG API returns gene associations in the format \code{ko:K00001\teco:b0001},
where the first column is the KO identifier and the second column is the
gene identifier (formatted as \code{species:gene_id}).</p>

        <h3 style="display: block">Authors</h3>

        xieguigang

        <h3>Value</h3>

        <p>Returns \code{NULL} invisibly.</p>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>kegg_api</em> version 0.1.0 <a
                href="../kegg_api.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="../../_assets/R_syntax.js"></script>
<script type="text/javascript" src="../../_assets/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>
