# annotation.genbank_kit

NCBI genbank assembly file I/O toolkit

+ [read.genbank](annotation.genbank_kit/read.genbank.1) read the given genbank assembly file.
+ [is.plasmid](annotation.genbank_kit/is.plasmid.1) 
+ [populate.genbank](annotation.genbank_kit/populate.genbank.1) populate a list of genbank data objects from a given list of files or stream.
+ [write.genbank](annotation.genbank_kit/write.genbank.1) save the modified genbank file
+ [as.genbank](annotation.genbank_kit/as.genbank.1) converts tabular data file to genbank assembly object
+ [feature](annotation.genbank_kit/feature.1) create new feature site
+ [add.feature](annotation.genbank_kit/add.feature.1) 
+ [enumerateFeatures](annotation.genbank_kit/enumerateFeatures.1) enumerate all features in the given NCBI genbank database object
+ [featureKeys](annotation.genbank_kit/featureKeys.1) 
+ [featureMeta](annotation.genbank_kit/featureMeta.1) 
+ [addMeta](annotation.genbank_kit/addMeta.1) 
+ [origin.fasta](annotation.genbank_kit/origin.fasta.1) get, add or replace the genome origin fasta sequence in the given genbank assembly file.
+ [getRNA.fasta](annotation.genbank_kit/getRNA.fasta.1) get all of the RNA gene its gene sequence in fasta sequence format.
+ [protein.fasta](annotation.genbank_kit/protein.fasta.1) get or set fasta sequence of all CDS feature in the given genbank assembly file.
+ [add.RNA.gene](annotation.genbank_kit/add.RNA.gene.1) 
