# annotation.genomics



+ [read.gtf](annotation.genomics/read.gtf.1) 
+ [read.gff](annotation.genomics/read.gff.1) 
+ [gff_features](annotation.genomics/gff_features.1) get gff features by id reference
+ [as.tabular](annotation.genomics/as.tabular.1) 
+ [as.geneTable](annotation.genomics/as.geneTable.1) 
+ [as.PTT](annotation.genomics/as.PTT.1) 
+ [upstream](annotation.genomics/upstream.1) Create the upstream location
+ [genes_features](annotation.genomics/genes_features.1) Extract all gene features from a given genomics context assembly data
+ [write.PTT_tabular](annotation.genomics/write.PTT_tabular.1) 
+ [operon_set](annotation.genomics/operon_set.1) load operon set data from the ODB database
+ [read.nucmer](annotation.genomics/read.nucmer.1) 
