<!DOCTYPE html><html lang="zh-CN">
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, minimum-scale=1.0, maximum-scale=1.0" />
    <title>GSVA</title>
    <meta name="author" content="xie.guigang@gcmodeller.org" />
    <meta name="copyright" content="SMRUCC genomics Copyright (c) 2020" />
    <meta name="keywords" content="GCModeller; Xanthomonas; Artificial Life" />
    <meta name="generator" content="https://github.com/xieguigang/xDoc" />
    <meta name="theme-color" content="#333" />
    <meta name="last-update" content="2023-01-04" />
    <meta name="description" content="" />
    <meta class="foundation-data-attribute-namespace" />
    <meta class="foundation-mq-xxlarge" />
    <meta class="foundation-mq-xlarge" />
    <meta class="foundation-mq-large" />
    <meta class="foundation-mq-medium" />
    <meta class="foundation-mq-small" />
    <meta class="foundation-mq-topbar" />
  </head>
  <body>
    <table width="100%" summary="page for {GSVA}">
      <tbody>
        <tr>
          <td>{GSVA}</td>
          <td style="text-align: right;">R# Documentation</td>
        </tr>
      </tbody>
    </table>
    <h1>GSVA</h1>
    <hr />
    <p>
      <code>
        <span style="color: blue;">imports</span>
        <span style="color: brown"> "GSVA"</span>
        <span style="color: blue;"> from</span>
        <span style="color: brown"> "gseakit"</span>;
                           </code>
    </p>
    <p></p>
    <div id="main-wrapper">
      <table>
        <tbody><tr>
  <td id="gsva">
    <a href="./GSVA/gsva.html">gsva</a>
  </td>
  <td><p>Gene Set Variation Analysis for microarray and RNA-seq data</p>

<p>Gene Set Variation Analysis (GSVA) is a non-parametric, unsupervised <br />
 method for estimating variation of gene set enrichment through the<br />
 samples of a expression data set. GSVA performs a change in coordinate<br />
 systems, transforming the data from a gene by sample matrix to a gene-set<br />
 by sample matrix, thereby allowing the evaluation of pathway enrichment <br />
 for each sample. This new matrix of GSVA enrichment scores facilitates<br />
 applying standard analytical methods like functional enrichment, <br />
 survival analysis, clustering, CNV-pathway analysis or cross-tissue <br />
 pathway analysis, in a pathway-centric manner.</p>

<p>main function of the package which estimates activity<br />
 scores For Each given gene-Set</p></td>
</tr>
<tr>
  <td id="diff">
    <a href="./GSVA/diff.html">diff</a>
  </td>
  <td><p>different analysis of the gsva result</p></td>
</tr>
<tr>
  <td id="matrix_to_diff">
    <a href="./GSVA/matrix_to_diff.html">matrix_to_diff</a>
  </td>
  <td><p>convert to diff data from dataframe</p></td>
</tr></tbody>
      </table>
    </div>
  </body>
</html>