<!DOCTYPE html><html lang="zh-CN">
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, minimum-scale=1.0, maximum-scale=1.0" />
    <title>GSVA</title>
    <meta name="author" content="xie.guigang@gcmodeller.org" />
    <meta name="copyright" content="SMRUCC genomics Copyright (c) 2022" />
    <meta name="keywords" content="R#; GSVA; gseakit" />
    <meta name="generator" content="https://github.com/rsharp-lang" />
    <meta name="theme-color" content="#333" />
    <meta name="description" content="Gene Set Variation Analysis for microarray and RNA-seq data..." />
    <meta class="foundation-data-attribute-namespace" />
    <meta class="foundation-mq-xxlarge" />
    <meta class="foundation-mq-xlarge" />
    <meta class="foundation-mq-large" />
    <meta class="foundation-mq-medium" />
    <meta class="foundation-mq-small" />
    <meta class="foundation-mq-topbar" />
    <style>

.table-three-line {
border-collapse:collapse; /* 关键属性：合并表格内外边框(其实表格边框有2px，外面1px，里面还有1px哦) */
border:solid #000000; /* 设置边框属性；样式(solid=实线)、颜色(#999=灰) */
border-width:2px 0 2px 0px; /* 设置边框状粗细：上 右 下 左 = 对应：1px 0 0 1px */
}
.left-1{
    border:solid #000000;border-width:1px 1px 2px 0px;padding:2px;
    font-weight:bolder;
}
.right-1{
    border:solid #000000;border-width:1px 0px 2px 1px;padding:2px;
    font-weight:bolder;
}
.mid-1{
    border:solid #000000;border-width:1px 1px 2px 1px;padding:2px;
    font-weight:bolder;
}
.left{
    border:solid #000000;border-width:1px 1px 1px 0px;padding:2px;
}
.right{
    border:solid #000000;border-width:1px 0px 1px 1px;padding:2px;
}
.mid{
    border:solid #000000;border-width:1px 1px 1px 1px;padding:2px;
}
table caption {font-size:14px;font-weight:bolder;}
</style>
  </head>
  <body>
    <table width="100%" summary="page for {GSVA}">
      <tbody>
        <tr>
          <td>{GSVA}</td>
          <td style="text-align: right;">R# Documentation</td>
        </tr>
      </tbody>
    </table>
    <h1>GSVA</h1>
    <hr />
    <p style="     font-size: 1.125em;     line-height: .8em;     margin-left: 0.5%;     background-color: #fbfbfb;     padding: 24px; ">
      <code>
        <span style="color: blue;">require</span>(<span style="color: black; font-weight: bold;">GCModeller</span>);
                               <br /><br /><span style="color: green;">#' Gene Set Variation Analysis for microarray and RNA-seq data</span><br /><span style="color: blue;">imports</span><span style="color: brown"> "GSVA"</span><span style="color: blue;"> from</span><span style="color: brown"> "gseakit"</span>;
                           </code>
    </p>
    <p><p>Gene Set Variation Analysis for microarray and RNA-seq data</p></p>
    <blockquote>
      <p style="font-style: italic; font-size: 0.9em;">
                           
                           </p>
    </blockquote>
    <div id="main-wrapper">
      <table class="table-three-line" style="display: none">
        <caption>.NET clr type export</caption>
        <tbody></tbody>
      </table>
      <br />
      <br />
      <table class="table-three-line">
        <caption>.NET clr function exports</caption>
        <tbody><tr>
  <td id="gsva">
    <a href="./GSVA/gsva.html">gsva</a>
  </td>
  <td><p>Gene Set Variation Analysis for microarray and RNA-seq data
 
 Gene Set Variation Analysis (GSVA) is a non-parametric, unsupervised 
 method for estimating variation of gene set enrichment through the
 samples of a expression data set. GSVA performs a change in coordinate
 systems, transforming the data from a gene by sample matrix to a gene-set
 by sample matrix, thereby allowing the evaluation of pathway enrichment 
 for each sample. This new matrix of GSVA enrichment scores facilitates
 applying standard analytical methods like functional enrichment, 
 survival analysis, clustering, CNV-pathway analysis or cross-tissue 
 pathway analysis, in a pathway-centric manner.
 
 main function of the package which estimates activity
 scores For Each given gene-Set</p></td>
</tr>
<tr>
  <td id="diff">
    <a href="./GSVA/diff.html">diff</a>
  </td>
  <td><p>different analysis of the gsva result</p></td>
</tr>
<tr>
  <td id="matrix_to_diff">
    <a href="./GSVA/matrix_to_diff.html">matrix_to_diff</a>
  </td>
  <td><p>convert to diff data from dataframe</p></td>
</tr></tbody>
      </table>
    </div>
    <hr />
    <div style="text-align: center;">[<a href="../index.html">Document Index</a>]</div>
  </body>
</html>
