<!DOCTYPE html>
<html>

<head>
    <title>do GSEA enrichment analysis</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="../../_assets/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>enrichment {GSEA}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>do GSEA enrichment analysis</h2>

        <h3>Description</h3>

        <p><br /> Gene set enrichment analysis (GSEA) (also called functional <br /> enrichment analysis or pathway enrichment analysis) is a <br /> method to identify classes of genes or proteins that are <br /> over-represented in a large set of genes or proteins, and <br /> may have an association with disease phenotypes. The method <br /> uses statistical approaches to identify significantly enriched <br /> or depleted groups of genes. Transcriptomics technologies <br /> and proteomics results often identify thousands of genes <br /> which are used for the analysis.</p>

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>enrichment</strong>(<i>background</i>, <i>geneSet</i>,
    <i>cut.size</i> = <span style='color: green;'>3</span>,
    <i>outputAll</i> = <span style='color: blue !important;'>TRUE</span>,
    <i>resize</i> = <span style='color: green;'>-1</span>,
    <i>showProgress</i> = <span style='color: blue !important;'>FALSE</span>,
    <i>...</i> = <span style='color: blue !important;'>NULL</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>background</dt>
<dd><p><p>a background model or metpa model</p></p></dd>


<dt>geneSet</dt>
<dd><p><p>a given geneset id list</p></p></dd>


<dt>args</dt>
<dd><p><p>the additional argument list that may be used, </p>

<p>for metpa background model, <code>topo</code> parameter will be used for <br />
 specific the network topology impact score source, value of the<br />
 <code>topo</code> parameter could be <code>dgr</code> or <code>rbc</code>.</p></p></dd>

        </table>


        <h3 style="display: none;">Details</h3>

        


        <h3>Value</h3>

        this function returns data object of type <a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/GSEA/EnrichmentResult.html">EnrichmentResult</a>.<h4>clr value class</h4><ul><li><a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/GSEA/EnrichmentResult.html">EnrichmentResult</a></li></ul>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>GSEA</em> version 1.0.0.0 <a
                href="../GSEA.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="../../_assets/R_syntax.js"></script>
<script type="text/javascript" src="../../_assets/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>