<!DOCTYPE html>
<html>

<head>
    <title>do GSEA enrichment analysis</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>enrichment {GSEA}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>do GSEA enrichment analysis</h2>

        <h3>Description</h3>

        <p><br /> Gene set enrichment analysis (GSEA) (also called functional <br /> enrichment analysis or pathway enrichment analysis) is a <br /> method to identify classes of genes or proteins that are <br /> over-represented in a large set of genes or proteins, and <br /> may have an association with disease phenotypes. The method <br /> uses statistical approaches to identify significantly enriched <br /> or depleted groups of genes. Transcriptomics technologies <br /> and proteomics results often identify thousands of genes <br /> which are used for the analysis.</p>

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>enrichment</strong>(<i>background</i>, <i>geneSet</i>,
    <i>expression</i> = <span style='color: blue !important;'>NULL</span>,
    <i>cut.size</i> = <span style='color: green;'>3</span>,
    <i>outputAll</i> = <span style='color: blue !important;'>TRUE</span>,
    <i>resize</i> = <span style='color: green;'>-1</span>,
    <i>showProgress</i> = <span style='color: blue !important;'>FALSE</span>,
    <i>...</i> = <span style='color: blue !important;'>NULL</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>background</dt>
<dd><p>a <a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/GSEA/Background.html">Background</a> model or @<code>T:SMRUCC.genomics.Analysis.HTS.GSEA.KnowledgeBase.Metabolism.Metpa.metpa</code> background model.</p></dd>


<dt>geneSet</dt>
<dd><p>a given geneset id list</p></dd>


<dt>expression</dt>
<dd><p>the expression value of the given gene set. this parameter could be missing. 
 default use the fisher test enrichment method for the given gene set when 
 the expression vector is missing. GSEA method will be applied for the enrichment 
 analysis if this parameter is not missing.</p></dd>


<dt>args</dt>
<dd><p>the additional argument list that may be used, 
 
 for metpa background model, <code>topo</code> parameter will be used for 
 specific the network topology impact score source, value of the
 <code>topo</code> parameter could be <code>dgr</code> or <code>rbc</code>. [as list]</p></dd>

        </table>


        <h3 style="display: block;">Details</h3>

        <p>the expression value could be the raw expression mean value, different expression 
 foldchange value for the group comparision, or other related gene set sort scores.</p>

        <h3 style="display: block">Authors</h3>

        gseakit

        <h3>Value</h3>

        this function returns data object of type <a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/GSEA/EnrichmentResult.html">EnrichmentResult</a>.<h4>clr value class</h4><ul><li><a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/GSEA/EnrichmentResult.html">EnrichmentResult</a></li></ul>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>GSEA</em> version 1.0.0.0 <a
                href="../GSEA.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>