# bioseq.fasta

Fasta sequence toolkit

+ [size](bioseq.fasta/size.1) get the sequence length
+ [chars](bioseq.fasta/chars.1) get alphabets represents of the fasta sequence
+ [mass](bioseq.fasta/mass.1) evaluate the molecule mass of the given sequence
+ [seq_formula](bioseq.fasta/seq_formula.1) 
+ [seq_sgt](bioseq.fasta/seq_sgt.1) Create algorithm for make sequence embedding
+ [seq_vector](bioseq.fasta/seq_vector.1) embedding the given fasta sequence as vector
+ [read.seq](bioseq.fasta/read.seq.1) Read a single fasta sequence file
+ [read.fasta](bioseq.fasta/read.fasta.1) read a fasta sequence collection file
+ [open.fasta](bioseq.fasta/open.fasta.1) open the fasta sequence file
+ [parse.fasta](bioseq.fasta/parse.fasta.1) parse the fasta sequence object from the given text data
+ [takes](bioseq.fasta/takes.1) 
+ [write.fasta](bioseq.fasta/write.fasta.1) write a fasta sequence or a collection of fasta sequence object
+ [translate](bioseq.fasta/translate.1) Do translation of the nt sequence to protein sequence
+ [MSA.of](bioseq.fasta/MSA.of.1) Do multiple sequence alignment
+ [as.fasta](bioseq.fasta/as.fasta.1) Create a fasta sequence collection object from any given sequence collection.
+ [fasta](bioseq.fasta/fasta.1) Create a new fasta sequence objects
+ [fasta.headers](bioseq.fasta/fasta.headers.1) get/set the fasta headers title
+ [fasta.titles](bioseq.fasta/fasta.titles.1) get the fasta titles from a collection of fasta sequence
+ [cut_seq.linear](bioseq.fasta/cut_seq.linear.1) cut part of the sequence
+ [open.fingerprint_writer](bioseq.fasta/open.fingerprint_writer.1) 
+ [write_fingerprint](bioseq.fasta/write_fingerprint.1) 
+ [read.fingerprint_bson](bioseq.fasta/read.fingerprint_bson.1) 
+ [make_clusterTree](bioseq.fasta/make_clusterTree.1) 
