<!DOCTYPE html>
<html>

<head>
    <title>find possible motifs of the given sequence collection</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>find_motifs {bioseq.patterns}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>find possible motifs of the given sequence collection</h2>

        <h3>Description</h3>

        

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>find_motifs</strong>(<i>fasta</i>,
    <i>minw</i> = <span style='color: green;'>8</span>,
    <i>maxw</i> = <span style='color: green;'>20</span>,
    <i>nmotifs</i> = <span style='color: green;'>-1</span>,
    <i>noccurs</i> = <span style='color: green;'>12</span>,
    <i>seedingCutoff</i> = <span style='color: green;'>0.65</span>,
    <i>scanMinW</i> = <span style='color: green;'>6</span>,
    <i>scanCutoff</i> = <span style='color: green;'>0.8</span>,
    <i>cleanMotif</i> = <span style='color: green;'>0.5</span>,
    <i>significant.sites</i> = <span style='color: green;'>4</span>,
    <i>seeds</i> = <span style='color: blue !important;'>NULL</span>,
    <i>debug</i> = <span style='color: blue !important;'>FALSE</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>fasta</dt>
<dd><p>should contains multiple sequence</p></dd>


<dt>minw%</dt>
<dd><p>-</p></dd>


<dt>maxw%</dt>
<dd><p>-</p></dd>


<dt>nmotifs</dt>
<dd><p>A number for limit the number of motif outputs:
 
 + negative integer/zero: no limits[default]
 + positive value: top motifs with score desc. [as integer]</p></dd>


<dt>noccurs%</dt>
<dd><p>-</p></dd>

        </table>


        <h3 style="display: none;">Details</h3>

        

        <h3 style="display: block">Authors</h3>

        seqtoolkit

        <h3>Value</h3>

        this function returns data object of type <a href="/vignettes/clr/SMRUCC/genomics/Analysis/SequenceTools/SequencePatterns/Motif/SequenceMotif.html">SequenceMotif</a>.<h4>clr value class</h4><ul><li><a href="/vignettes/clr/SMRUCC/genomics/Analysis/SequenceTools/SequencePatterns/Motif/SequenceMotif.html">SequenceMotif</a></li></ul>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>bioseq.patterns</em> version 1.0.0.0 <a
                href="../bioseq.patterns.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>