<!DOCTYPE html>
<html>

<head>
    <title>calculate average value of the gene expression for</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>average {geneExpression}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>calculate average value of the gene expression for</h2>

        <h3>Description</h3>

        <p>each sample group.<br /> <br /> this method can be apply for reduce data size when <br /> create some plot for visualize the gene expression<br /> patterns across the sample groups.</p>

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>average</strong>(<i>matrix</i>,
    <i>sampleinfo</i> = <span style='color: blue !important;'>NULL</span>,
    <i>strict</i> = <span style='color: blue !important;'>TRUE</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>matrix</dt>
<dd><p>a gene expression matrix. [as <a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/DataFrame/Matrix.html">Matrix</a>]</p></dd>


<dt>sampleinfo</dt>
<dd><p>The sample group data. [as <a href="/vignettes/clr/SMRUCC/genomics/GCModeller/Workbench/ExperimentDesigner/SampleInfo.html">SampleInfo</a>]</p></dd>


<dt>strict</dt>
<dd><p>will try to ignores of the missing sample if strict option is off. [as boolean]</p></dd>

        </table>


        <h3 style="display: none;">Details</h3>

        

        <h3 style="display: block">Authors</h3>

        phenotype_kit

        <h3>Value</h3>

        <p>this function return value is determined based on the sampleinfo parameter:
 
 1. for sampleinfo not nothing, a matrix with sample group as the sample feature data will be returns
 2. for missing sampleinfo data, a numeric vector of average value for each gene feature will be returns</p><h4>clr value class</h4><ul><li><a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/DataFrame/Matrix.html">Matrix</a></li><li>double</li></ul>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>geneExpression</em> version 1.0.0.0 <a
                href="../geneExpression.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>
