<!DOCTYPE html>
<html>

<head>
    <title>Z-score normalized of the expression data matrix</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>z_score {geneExpression}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>Z-score normalized of the expression data matrix</h2>

        <h3>Description</h3>

        <p><br /> To avoid the influence of expression level to the <br /> clustering analysis, z-score transformation can <br /> be applied to covert the expression values to <br /> z-scores by performing the following formula:<br /> <br /> <code>&lt;br /> z = (x - u) / sd&lt;br /> </code>`<br /> <br /> x is value to be converted (e.g., a expression value <br /> of a genomic feature in one condition), µ is the <br /> population mean (e.g., average expression value Of <br /> a genomic feature In different conditions), σ Is the <br /> standard deviation (e.g., standard deviation of <br /> expression of a genomic feature in different conditions).</p>

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>z_score</strong>(<i>x</i>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>x</dt>
<dd><p>a gene expression matrix. [as <a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/DataFrame/Matrix.html">Matrix</a>]</p></dd>

        </table>


        <h3 style="display: block;">Details</h3>

        <h4>Standard score(z-score)</h4>

<p>In statistics, the standard score is the signed number of standard deviations by which the value of 
 an observation or data point is above the mean value of what is being observed or measured. Observed 
 values above the mean have positive standard scores, while values below the mean have negative 
 standard scores. The standard score is a dimensionless quantity obtained by subtracting the population 
 mean from an individual raw score and then dividing the difference by the population standard deviation. 
 This conversion process is called standardizing or normalizing (however, "normalizing" can refer to 
 many types of ratios; see normalization for more).
 
 ><a href='https://en.wikipedia.org/wiki/Standard<em>score' title='https://en.wikipedia.org/wiki/Standard</em>score'>https://en.wikipedia.org/wiki/Standard_score</a></p>

        <h3 style="display: block">Authors</h3>

        phenotype_kit

        <h3>Value</h3>

        <p>the HTS matrix object has been normalized in each gene 
 expression row, z-score is calculated for each gene row
 across multiple sample expression data.</p><h4>clr value class</h4><ul><li><a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/DataFrame/Matrix.html">Matrix</a></li></ul>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>geneExpression</em> version 1.0.0.0 <a
                href="../geneExpression.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>