# kmers



+ [write.kmers_background](kmers/write.kmers_background.1) 
+ [read.kmers_background](kmers/read.kmers_background.1) 
+ [bayes_background](kmers/bayes_background.1) 
+ [read_seqid](kmers/read_seqid.1) 
+ [bloom_filters](kmers/bloom_filters.1) 
+ [make_classify](kmers/make_classify.1) just make reads classify of the fastq reads based on the k-mer distribution
+ [bayes_estimate](kmers/bayes_estimate.1) 
+ [bayes_abundance](kmers/bayes_abundance.1) quantify of the metagenome community via kmers and bayes method
+ [as.bloom_filter](kmers/as.bloom_filter.1) cast the genomics sequence as kmer based bloom filter model
+ [parse_kraken_output](kmers/parse_kraken_output.1) 
+ [parse_kraken_report](kmers/parse_kraken_report.1) 
+ [read.kraken2](kmers/read.kraken2.1) 
+ [read_brackens](kmers/read_brackens.1) 
+ [filter_classification](kmers/filter_classification.1) 
+ [filter_reads](kmers/filter_reads.1) 
+ [seq_info](kmers/seq_info.1) 
+ [kraken_seqs](kmers/kraken_seqs.1) extract gene/genomics sequences from genbank file for kraken2 sequence classification
