<!DOCTYPE html>
<html>

<head>
    <title>export gene fasta from the given genbank assembly file</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>export_geneNt_fasta {GenBank}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>export gene fasta from the given genbank assembly file</h2>

        <h3>Description</h3>

        

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>export_geneNt_fasta</strong>(<i>gb</i>,
    <i>title</i> = <span style='color: brown;'><strong>"<gb_asm_id>.<locus_tag> <nucl_loc> <product>|<lineage>"</strong></span>,
    <i>key</i> = [<span style='color: brown;'><strong>'CDS'</strong></span>],
    <i>unique.names</i> = <span style='color: blue !important;'>FALSE</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>gb</dt>
<dd><p>[as <a href="/vignettes/clr/SMRUCC/genomics/Assembly/NCBI/GenBank/GBFF/File.html">File</a>]</p></dd>


<dt>title</dt>
<dd><p>[as string]</p></dd>


<dt>unique.names</dt>
<dd><p>processing the possible duplicated header as unique id, this option is usually when you use this function for build a sequence database for salmon tool. [as boolean]</p></dd>

        </table>


        <h3 style="display: block;">Details</h3>

        <p>fasta title is build with a string template, there are some reserved template keyword for this function:
 
 1. ncbi_taxid - is the ncbi taxonomy id that extract from the genbank assembly
 2. lineage - taxonomy lineage in biom style string, which is extract from the genbank assembly its source information
 3. gb<em>asm</em>id - the ncbi accession id of the genbank assembly
 4. nucl_loc - the nucleotide sequence location on the genomics sequence
 
 ##### about the salmon duplicated id error
 
 if you encounter this error while build sequence index by using salmon tool, please set the <code>unique.names</code> parameter to TRUE


<pre><code class=""> counted k-mers for 110000 transcripts[2026-01-06 15:08:42.160] [puff::index::jointLog] [error] In FixFasta, two references with the same name but different sequences: AM295250.SCA_1840. We require that all input records have a unique name up to the first whitespace (or user-provided separator) character.
</code></pre>

</p>

        <h3 style="display: block">Authors</h3>

        seqtoolkit

        <h3>Value</h3>

        this function returns data object of type <a href="/vignettes/clr/SMRUCC/genomics/SequenceModel/FASTA/FastaFile.html">FastaFile</a>.<h4>clr value class</h4><ul><li><a href="/vignettes/clr/SMRUCC/genomics/SequenceModel/FASTA/FastaFile.html">FastaFile</a></li></ul>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>GenBank</em> version 1.0.0.0 <a
                href="../GenBank.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>
