<!DOCTYPE html>
<html>

<head>
    <title>SMRUCC.genomics.Assembly.KEGG.DBGET.bGetObject.Module</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>Module {SMRUCC.genomics.Assembly.KEGG.DBGET.bGetObject}</td>
                <td style="text-align: right;">.NET clr documentation</td>
            </tr>
        </table>

        <h2>Module</h2>

        <h3>Description</h3>

        <p>KEGG MODULE is a collection of manually defined functional units, called KEGG modules and identified by the M numbers, <br />
 used for annotation and biological interpretation of sequenced genomes. </p>

<p>There are four types of KEGG modules:</p>

<ul>
<li><code>pathway modules</code>      – representing tight functional units In KEGG metabolic pathway maps, such As M00002 (Glycolysis, <br />
core Module involving three-carbon compounds)</li>
<li><code>structural complexes</code> – often forming molecular machineries, such As M00072 (Oligosaccharyltransferase)</li>
<li><code>functional sets</code>      – For other types Of essential sets, such As M00360 (Aminoacyl-tRNA synthases, prokaryotes)</li>
<li><p><code>signature modules</code>    – As markers Of phenotypes, such As M00363 (EHEC pathogenicity signature, Shiga toxin)</p>

<p>The entire list Of KEGG modules can be viewed from the BRITE hierarchy file:<br />
KEGG modules</p>

<p>Logical Expression</p>

<p>The M number entry Is defined by a logical expression Of K numbers (And other M numbers), allowing automatic evaluation <br />
Of whether the gene Set Is complete, i.e., the Module Is present, In a given genome. A space Or a plus sign represents <br />
an And operation, And a comma sign represents an Or operation In this expression. A plus sign Is used For a molecular <br />
complex And a minus sign designates an Optional item In the complex. </p>

<p>Each space-separated unit Is called a block, And the distinction Is made for:<br />
complete modules<br />
incomplete but almost complete modules With only 1 Or 2 blocks missing<br />
all modules that contain any matching K numbers<br />
in the KEGG MODULE database, as well as in the KEGG Mapper tool "Reconstruct Module".<br />
Module Map</p>

<p>KEGG modules are associated With graphical diagrams called Module maps. For example, M00002 represents glycolysis core <br />
Module involving three-carbon compounds And its organism specific Module can be selected from the pop-up menu Or <br />
directly specified In the form Of hsa_M00002. While KEGG pathway maps are all manually drawn, KEGG Module maps are <br />
computationally generated from the text definition describe above. There Is also a difference Of how organism-specific <br />
versions are generated. Organism specific Module entries are created only For complete modules, While organism-specific <br />
pathway map entries are created When a few matching elements exist under a predefined taxonomic constraint.<br />
Ortholog Table</p>

<p>The ortholog table Is a useful tool To check completeness And consistency Of genome annotations. It shows currently <br />
annotated genes In individual genomes For a given Set Of K numbers, together With coloring Of adjacent genes <br />
(operon-Like structures) On the chromosome. Each KEGG Module contains a link To the corresponding ortholog table, <br />
such As For M00165, together With Option To Select complete Or other modules.<br />
Taxonomy Mapping</p>

<p>Each KEGG module also contains a link to KEGG taxonomy mapping, showing which organisms Or organism groups have the module <br />
under the taxonomic classification of KEGG organisms. The taxonomy link from the ortholog table (designated by T) allows <br />
mapping of both complete And incomplete modules. The result Is shown in the color coding shown below.<br />
complete<br />
incomplete, 1 block missing<br />
incomplete, 2 blocks missing<br />
KEGG Modules And Reaction Modules</p>

<p>It Is interesting to note that KEGG modules sometimes correspond to reaction modules extracted from purely chemical <br />
properties as summarized in the following BRITE hierarchy file:<br />
KEGG reaction modules<br />
A New category of KEGG metabolic pathway maps, called overview maps, shows this correspondence as well as an overall <br />
architecture of the metabolic network. The following Is an example taken from the overview map for Degradation of <br />
aromatic compounds. </p>

<p>A single M number Or a combination of M numbers can be used for characterizing phenotypic features encoded in the genome. <br />
For example, the BTX (benzene, toluene, And xylene) degradation capacity can be seen from the following diagram where <br />
M numbers are linked to the ortholog tables indicating which organisms have complete modules.<br />
benzene    — M00548 →  catechol<br />
toluene    — M00538 →  benzoate    — M00551 →  catechol    — M00569 →  meta-cleavage<br />
— M00568 → ortho-cleavage<br />
xylene — M00537 →  methyl-<br />
benzoate   — M00551 →  methyl-<br />
catechol   — M00569 →  meta-cleavage<br />
— M00568 → ortho-cleavage<br />
This example can be rewritten In terms Of the reaction modules.<br />
benzene    — RM006 →   catechol<br />
toluene    — RM003 →   benzoate    — RM005 →   catechol    — RM009 →   meta-cleavage<br />
— RM008 →  ortho-cleavage<br />
xylene — RM003 →   methyl-<br />
benzoate   — RM005 →   methyl-<br />
catechol   — RM009 →   meta-cleavage<br />
— RM008 →  ortho-cleavage<br />
See more details In:<br />
Muto, A., Kotera, M., Tokimatsu, T., Nakagawa, Z., Goto, S., And Kanehisa, M.; Modular architecture of metabolic <br />
pathways revealed by conserved sequences of reactions. J. Chem. Inf. Model. 53, 613-622 (2013). [pubmed] [pdf] </p>

<p>Kanehisa, M.; Chemical And genomic evolution of enzyme-catalyzed reaction networks. FEBS Lett. 587, 2731-2737 (2013). [pubmed] [pdf] </p>

<p>Kanehisa, M., Goto, S., Sato, Y., Kawashima, M., Furumichi, M., And Tanabe, M.; Data, information, knowledge And <br />
principle back to metabolism in KEGG. Nucleic Acids Res. 42, D199–D205 (2014). [pubmed] [pdf]</p></li>
</ul>

        <h3>Declare</h3>        

        <pre>
            <code id="clr_ts">
# namespace SMRUCC.genomics.Assembly.KEGG.DBGET.bGetObject
export class Module extends <a href="/vignettes/clr/SMRUCC/genomics/ComponentModel/Annotation/PathwayBrief.html">PathwayBrief</a> {
   briteID: string;
   compound: <a href="/vignettes/clr/Microsoft/VisualBasic/Text/Xml/Models/NamedValue.html">NamedValue[]</a>;
   description: string;
   EntryId: string;
   name: string;
   pathway: <a href="/vignettes/clr/Microsoft/VisualBasic/Text/Xml/Models/NamedValue.html">NamedValue[]</a>;
   pathwayGenes: <a href="/vignettes/clr/Microsoft/VisualBasic/Text/Xml/Models/NamedValue.html">NamedValue[]</a>;
   reaction: <a href="/vignettes/clr/Microsoft/VisualBasic/Text/Xml/Models/NamedValue.html">NamedValue[]</a>;
   Stylesheet: <a href="/vignettes/clr/Microsoft/VisualBasic/ComponentModel/XmlStyleProcessor.html">XmlStyleProcessor</a>;
   TypeComment: <a href="/vignettes/clr/System/Xml/XmlComment.html">XmlComment</a>;
}
</code>
        </pre>

        <h3>.NET clr type reference tree</h3>

        <ol>
<li>this class extends from <a href="/vignettes/clr/SMRUCC/genomics/ComponentModel/Annotation/PathwayBrief.html">PathwayBrief</a> class: <a href="/vignettes/clr/SMRUCC/genomics/ComponentModel/Annotation/PathwayBrief.html">PathwayBrief</a></li>
<li>use by <i>property</i> member <code>compound</code>: <a href="/vignettes/clr/Microsoft/VisualBasic/Text/Xml/Models/NamedValue.html">NamedValue</a></li>
<li>use by <i>property</i> member <code>pathway</code>: <a href="/vignettes/clr/Microsoft/VisualBasic/Text/Xml/Models/NamedValue.html">NamedValue</a></li>
<li>use by <i>property</i> member <code>pathwayGenes</code>: <a href="/vignettes/clr/Microsoft/VisualBasic/Text/Xml/Models/NamedValue.html">NamedValue</a></li>
<li>use by <i>property</i> member <code>reaction</code>: <a href="/vignettes/clr/Microsoft/VisualBasic/Text/Xml/Models/NamedValue.html">NamedValue</a></li>
<li>use by <i>property</i> member <code>Stylesheet</code>: <a href="/vignettes/clr/Microsoft/VisualBasic/ComponentModel/XmlStyleProcessor.html">XmlStyleProcessor</a></li>
<li>use by <i>property</i> member <code>TypeComment</code>: <a href="/vignettes/clr/System/Xml/XmlComment.html">XmlComment</a></li>
</ol>


        <hr />
        <div style="text-align: center;">[Package <em>{$package}</em> version {$version} <a
                href="../{$package}.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("clr_ts");</script>

</html>