<!DOCTYPE html>
<html>

<head>
    <title>cut part of the sequence</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>cut_seq.linear {bioseq.fasta}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>cut part of the sequence</h2>

        <h3>Description</h3>

        

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>cut_seq.linear</strong>(<i>seq</i>, <i>loci</i>,
    <i>nt.auto.reverse</i> = <span style='color: blue !important;'>FALSE</span>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>seq</dt>
<dd><p>-</p></dd>


<dt>loci</dt>
<dd><p>the location region data for make cut of the sequence site, data model could be:
 
 1. for nucleotide sequence, <a href="/vignettes/clr/SMRUCC/genomics/ComponentModel/Loci/NucleotideLocation.html">NucleotideLocation</a> should be used,
 2. for general sequence data, <a href="/vignettes/clr/SMRUCC/genomics/ComponentModel/Loci/Location.html">Location</a> should be used.</p></dd>


<dt>nt.auto.reverse</dt>
<dd><p>make auto reverse of the nucleotide sequence if the given location is on 
 the <a href="/vignettes/clr/SMRUCC/genomics/ComponentModel/Loci/Strands.html">Strands</a>.<a href="/vignettes/clr/SMRUCC/genomics/ComponentModel/Loci/Strands.html">Reverse</a> direction. [as boolean]</p></dd>


<dt>env</dt>
<dd><p>[as <a href="/vignettes/clr/SMRUCC/Rsharp/Runtime/Environment.html">Environment</a>]</p></dd>

        </table>


        <h3 style="display: none;">Details</h3>

        

        <h3 style="display: block">Authors</h3>

        seqtoolkit

        <h3>Value</h3>

        this function returns data object of type <i>any</i> kind.<h4>clr value class</h4><ul><li><i>any</i> kind</li></ul>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>bioseq.fasta</em> version 1.0.0.0 <a
                href="../bioseq.fasta.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>
