<!DOCTYPE html><html lang="zh-CN">
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, minimum-scale=1.0, maximum-scale=1.0" />
    <title>bioseq.patterns</title>
    <meta name="author" content="xie.guigang@gcmodeller.org" />
    <meta name="copyright" content="SMRUCC genomics Copyright (c) 2022" />
    <meta name="keywords" content="R#; bioseq.patterns; seqtoolkit" />
    <meta name="generator" content="https://github.com/rsharp-lang" />
    <meta name="theme-color" content="#333" />
    <meta name="description" content="Tools for sequence patterns..." />
    <meta class="foundation-data-attribute-namespace" />
    <meta class="foundation-mq-xxlarge" />
    <meta class="foundation-mq-xlarge" />
    <meta class="foundation-mq-large" />
    <meta class="foundation-mq-medium" />
    <meta class="foundation-mq-small" />
    <meta class="foundation-mq-topbar" />
    <style>

.table-three-line {
border-collapse:collapse; /* 关键属性：合并表格内外边框(其实表格边框有2px，外面1px，里面还有1px哦) */
border:solid #000000; /* 设置边框属性；样式(solid=实线)、颜色(#999=灰) */
border-width:2px 0 2px 0px; /* 设置边框状粗细：上 右 下 左 = 对应：1px 0 0 1px */
}
.left-1{
    border:solid #000000;border-width:1px 1px 2px 0px;padding:2px;
    font-weight:bolder;
}
.right-1{
    border:solid #000000;border-width:1px 0px 2px 1px;padding:2px;
    font-weight:bolder;
}
.mid-1{
    border:solid #000000;border-width:1px 1px 2px 1px;padding:2px;
    font-weight:bolder;
}
.left{
    border:solid #000000;border-width:1px 1px 1px 0px;padding:2px;
}
.right{
    border:solid #000000;border-width:1px 0px 1px 1px;padding:2px;
}
.mid{
    border:solid #000000;border-width:1px 1px 1px 1px;padding:2px;
}
table caption {font-size:14px;font-weight:bolder;}
</style>
  </head>
  <body>
    <table width="100%" summary="page for {bioseq.patterns}">
      <tbody>
        <tr>
          <td>{bioseq.patterns}</td>
          <td style="text-align: right;">R# Documentation</td>
        </tr>
      </tbody>
    </table>
    <h1>bioseq.patterns</h1>
    <hr />
    <p style="     font-size: 1.125em;     line-height: .8em;     margin-left: 0.5%;     background-color: #fbfbfb;     padding: 24px; ">
      <code>
        <span style="color: blue;">require</span>(<span style="color: black; font-weight: bold;">GCModeller</span>);
                               <br /><br /><span style="color: green;">#' Tools for sequence patterns</span><br /><span style="color: blue;">imports</span><span style="color: brown"> "bioseq.patterns"</span><span style="color: blue;"> from</span><span style="color: brown"> "seqtoolkit"</span>;
                           </code>
    </p>
    <p><p>Tools for sequence patterns</p></p>
    <blockquote>
      <p style="font-style: italic; font-size: 0.9em;">
                           
                           </p>
    </blockquote>
    <div id="main-wrapper">
      <table class="table-three-line">
        <tbody><tr>
  <td id="view.sites">
    <a href="./bioseq.patterns/view.sites.html">view.sites</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="read.motifs">
    <a href="./bioseq.patterns/read.motifs.html">read.motifs</a>
  </td>
  <td><p>read sequence motif json file.</p></td>
</tr>
<tr>
  <td id="read.scans">
    <a href="./bioseq.patterns/read.scans.html">read.scans</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="as.seq_graph">
    <a href="./bioseq.patterns/as.seq_graph.html">as.seq_graph</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="motif.find_sites">
    <a href="./bioseq.patterns/motif.find_sites.html">motif.find_sites</a>
  </td>
  <td><p>Find target loci site based on the given motif model</p></td>
</tr>
<tr>
  <td id="palindrome.mirror">
    <a href="./bioseq.patterns/palindrome.mirror.html">palindrome.mirror</a>
  </td>
  <td><p>Search mirror palindrome sites for a given seed sequence</p></td>
</tr>
<tr>
  <td id="seeds">
    <a href="./bioseq.patterns/seeds.html">seeds</a>
  </td>
  <td><p>Create seeds</p></td>
</tr>
<tr>
  <td id="motifString">
    <a href="./bioseq.patterns/motifString.html">motifString</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="find_motifs">
    <a href="./bioseq.patterns/find_motifs.html">find_motifs</a>
  </td>
  <td><p>find possible motifs of the given sequence collection</p></td>
</tr>
<tr>
  <td id="plot.seqLogo">
    <a href="./bioseq.patterns/plot.seqLogo.html">plot.seqLogo</a>
  </td>
  <td><p>Drawing the sequence logo just simply modelling this motif site <br />
 from the clustal multiple sequence alignment.</p></td>
</tr>
<tr>
  <td id="scaffold.orthogonality">
    <a href="./bioseq.patterns/scaffold.orthogonality.html">scaffold.orthogonality</a>
  </td>
  <td><p>analyses orthogonality of two DNA-Origami scaffold strands.<br />
 Multiple criteria For orthogonality Of the two sequences can be specified<br />
 to determine the level of orthogonality.</p></td>
</tr></tbody>
      </table>
    </div>
  </body>
</html>