<!DOCTYPE html><html lang="zh-CN">
  <head>
    <meta http-equiv="content-type" content="text/html; charset=UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, minimum-scale=1.0, maximum-scale=1.0" />
    <title>OTU_table</title>
    <meta name="author" content="xie.guigang@gcmodeller.org" />
    <meta name="copyright" content="SMRUCC genomics Copyright (c) 2022" />
    <meta name="keywords" content="R#; OTU_table; metagenomics_kit" />
    <meta name="generator" content="https://github.com/rsharp-lang" />
    <meta name="theme-color" content="#333" />
    <meta name="description" content="Tools for handling OTU table data..." />
    <meta class="foundation-data-attribute-namespace" />
    <meta class="foundation-mq-xxlarge" />
    <meta class="foundation-mq-xlarge" />
    <meta class="foundation-mq-large" />
    <meta class="foundation-mq-medium" />
    <meta class="foundation-mq-small" />
    <meta class="foundation-mq-topbar" />
    <style>

.table-three-line {
border-collapse:collapse; /* 关键属性：合并表格内外边框(其实表格边框有2px，外面1px，里面还有1px哦) */
border:solid #000000; /* 设置边框属性；样式(solid=实线)、颜色(#999=灰) */
border-width:2px 0 2px 0px; /* 设置边框状粗细：上 右 下 左 = 对应：1px 0 0 1px */
}
.left-1{
    border:solid #000000;border-width:1px 1px 2px 0px;padding:2px;
    font-weight:bolder;
}
.right-1{
    border:solid #000000;border-width:1px 0px 2px 1px;padding:2px;
    font-weight:bolder;
}
.mid-1{
    border:solid #000000;border-width:1px 1px 2px 1px;padding:2px;
    font-weight:bolder;
}
.left{
    border:solid #000000;border-width:1px 1px 1px 0px;padding:2px;
}
.right{
    border:solid #000000;border-width:1px 0px 1px 1px;padding:2px;
}
.mid{
    border:solid #000000;border-width:1px 1px 1px 1px;padding:2px;
}
table caption {font-size:14px;font-weight:bolder;}
</style>
  </head>
  <body>
    <table width="100%" summary="page for {OTU_table}">
      <tbody>
        <tr>
          <td>{OTU_table}</td>
          <td style="text-align: right;">R# Documentation</td>
        </tr>
      </tbody>
    </table>
    <h1>OTU_table</h1>
    <hr />
    <p style="     font-size: 1.125em;     line-height: .8em;     margin-left: 0.5%;     background-color: #fbfbfb;     padding: 24px; ">
      <code>
        <span style="color: blue;">require</span>(<span style="color: black; font-weight: bold;">GCModeller</span>);
                               <br /><br /><span style="color: green;">#' Tools for handling OTU table data</span><br /><span style="color: blue;">imports</span><span style="color: brown"> "OTU_table"</span><span style="color: blue;"> from</span><span style="color: brown"> "metagenomics_kit"</span>;
                           </code>
    </p>
    <p><p>Tools for handling OTU table data</p></p>
    <blockquote>
      <p style="font-style: italic; font-size: 0.9em;">
                           <h3>Operational taxonomic unit (OTU)</h3>

<p>OTU's are used to categorize bacteria based on sequence similarity.
 
 In 16S metagenomics approaches, OTUs are cluster of similar sequence variants of the 
 16S rDNA marker gene sequence. Each of these cluster is intended to represent a 
 taxonomic unit of a bacteria species or genus depending on the sequence similarity 
 threshold. Typically, OTU cluster are defined by a 97% identity threshold of the 16S 
 gene sequences to distinguish bacteria at the genus level.</p>

<p>Species separation requires a higher threshold Of 98% Or 99% sequence identity, Or 
 even better the use Of exact amplicon sequence variants (ASV) instead Of OTU sequence 
 clusters.</p>
                           </p>
    </blockquote>
    <div id="main-wrapper">
      <table class="table-three-line" style="display: block">
        <caption>.NET clr type export</caption>
        <tbody><tr>
  <td id="OTU_table">
    <a href="/vignettes/clr/SMRUCC/genomics/Analysis/Metagenome/OTUTable.html">OTU_table: OTUTable</a>
  </td>
  <td><h3>OTU table (sequence count table)</h3>

<p>A OTU table contains the number of sequences that are observed for each taxonomic 
 unit (OTUs) in each samples. Columns usually represent samples and rows represent
 genera or species specific taxonomic units (OTUs). OTU tables are often saved as
 BIOM formatted files.
 
 ### Limited taxonomic resolution
 
 OTU resolution depends On the 16S approach which has some limits In distinguishing at 
 the species level, For example,
 
 Escherichia coli And Shigella spp. share almost identical 16S rRNA gene sequences.
 
 Alternative approaches are developed To achieve higher resolution up To strain level 
 by considering larger Or complete sets Of genes.</p></td>
</tr></tbody>
      </table>
      <br />
      <br />
      <table class="table-three-line">
        <caption>.NET clr function exports</caption>
        <tbody><tr>
  <td id="as.data.frame.otutable">
    <a href="./OTU_table/as.data.frame.otutable.html">as.data.frame.otutable</a>
  </td>
  <td><a href="/vignettes/clr/SMRUCC/genomics/Analysis/Metagenome/OTUTable.html">OTUTable</a>: <h3>OTU table (sequence count table)</h3>

<p>A OTU table contains the number of sequences that are observed for each taxonomic 
 unit (OTUs) in each samples. Columns usually represent samples and rows represent
 genera or species specific taxonomic units (OTUs). OTU tables are often saved as
 BIOM formatted files.
 
 ### Limited taxonomic resolution
 
 OTU resolution depends On the 16S approach which has some limits In distinguishing at 
 the species level, For example,
 
 Escherichia coli And Shigella spp. share almost identical 16S rRNA gene sequences.
 
 Alternative approaches are developed To achieve higher resolution up To strain level 
 by considering larger Or complete sets Of genes.</p></td>
</tr>
<tr>
  <td id="relative_abundance">
    <a href="./OTU_table/relative_abundance.html">relative_abundance</a>
  </td>
  <td><p>Transform abundance data in an otu_table to relative abundance, sample-by-sample. 
 
 Transform abundance data into relative abundance, i.e. proportional data. This is 
 an alternative method of normalization and may not be appropriate for all datasets,
 particularly if your sequencing depth varies between samples.</p></td>
</tr>
<tr>
  <td id="median_scale">
    <a href="./OTU_table/median_scale.html">median_scale</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="impute_missing">
    <a href="./OTU_table/impute_missing.html">impute_missing</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="filter">
    <a href="./OTU_table/filter.html">filter</a>
  </td>
  <td><p>filter the otu data which has relative abundance greater than the given threshold</p></td>
</tr>
<tr>
  <td id="read.OTUtable">
    <a href="./OTU_table/read.OTUtable.html">read.OTUtable</a>
  </td>
  <td><p>read 16s OTU table</p></td>
</tr>
<tr>
  <td id="read.OTUdata">
    <a href="./OTU_table/read.OTUdata.html">read.OTUdata</a>
  </td>
  <td></td>
</tr>
<tr>
  <td id="batch_combine">
    <a href="./OTU_table/batch_combine.html">batch_combine</a>
  </td>
  <td><p>combine of two batch data</p></td>
</tr>
<tr>
  <td id="otu_from_matrix">
    <a href="./OTU_table/otu_from_matrix.html">otu_from_matrix</a>
  </td>
  <td><p>cast the expression matrix to the otu data</p></td>
</tr>
<tr>
  <td id="as.hts_matrix">
    <a href="./OTU_table/as.hts_matrix.html">as.hts_matrix</a>
  </td>
  <td><p>Create expression matrix data from a given otu table</p></td>
</tr>
<tr>
  <td id="as.OTU_table">
    <a href="./OTU_table/as.OTU_table.html">as.OTU_table</a>
  </td>
  <td><p>convert the mothur rank tree as the OTU table</p></td>
</tr>
<tr>
  <td id="make_otu_table">
    <a href="./OTU_table/make_otu_table.html">make_otu_table</a>
  </td>
  <td></td>
</tr></tbody>
      </table>
    </div>
    <hr />
    <div style="text-align: center;">[<a href="../index.html">Document Index</a>]</div>
  </body>
</html>
