<!DOCTYPE html>
<html>

<head>
    <title>Gene Set Variation Analysis for microarray and RNA-seq data</title>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
    <link rel="stylesheet" type="text/css" href="https://gcmodeller.org/lib/page.css">
</head>

<body>
    <div class="container">

        <table style="width: 100%;">
            <tr>
                <td>gsva {GSVA}</td>
                <td style="text-align: right;">R Documentation</td>
            </tr>
        </table>

        <h2>Gene Set Variation Analysis for microarray and RNA-seq data</h2>

        <h3>Description</h3>

        <p><br /> Gene Set Variation Analysis (GSVA) is a non-parametric, unsupervised <br /> method for estimating variation of gene set enrichment through the<br /> samples of a expression data set. GSVA performs a change in coordinate<br /> systems, transforming the data from a gene by sample matrix to a gene-set<br /> by sample matrix, thereby allowing the evaluation of pathway enrichment <br /> for each sample. This new matrix of GSVA enrichment scores facilitates<br /> applying standard analytical methods like functional enrichment, <br /> survival analysis, clustering, CNV-pathway analysis or cross-tissue <br /> pathway analysis, in a pathway-centric manner.<br /> <br /> main function of the package which estimates activity<br /> scores For Each given gene-Set</p>

        <h3>Usage</h3>

        <div style="width: 100%; padding-left: 3%; color: black;">
            <pre><code class='language-R' style="color: black; font-weight:normal;"><strong>gsva</strong>(<i>expr</i>, <i>geneSet</i>);</code></pre>
        </div>

        <h3>Arguments</h3>

        <table>
            
<dt>expr</dt>
<dd><p>A raw gene expression data matrix object</p></dd>


<dt>geneSet</dt>
<dd><p>A gsea enrichment <a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/GSEA/Background.html">Background</a> model</p></dd>


<dt>env</dt>
<dd><p>[as <a href="/vignettes/clr/SMRUCC/Rsharp/Runtime/Environment.html">Environment</a>]</p></dd>

        </table>


        <h3 style="display: block;">Details</h3>

        <p>HÃ¤nzelmann S., Castelo R. and Guinney J. GSVA: gene set variation analysis
 for microarray and RNA-Seq data. BMC Bioinformatics, 14:7, 2013.</p>

        <h3 style="display: block">Authors</h3>

        gseakit

        <h3>Value</h3>

        this function returns data object of type <a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/DataFrame/Matrix.html">Matrix</a>.<h4>clr value class</h4><ul><li><a href="/vignettes/clr/SMRUCC/genomics/Analysis/HTS/DataFrame/Matrix.html">Matrix</a></li></ul>

        <h3 style="display: none;">Examples</h3>

        

        <hr />
        <div style="text-align: center;">[Package <em>GSVA</em> version 1.0.0.0 <a
                href="../GSVA.html">Index</a>]
        </div>
    </div>
</body>

<script type="text/javascript" src="https://gcmodeller.org/lib/R_syntax.js"></script>
<script type="text/javascript" src="https://gcmodeller.org/lib/highlights.js"></script>
<script type="text/javascript">r_highlights("example_r");</script>

</html>
